pkg/routeagent_driver/handlers/kubeproxy/routes_iface.go

```diff
@@ -153,161 +154,123 @@ func (kp *SyncHandler) configureRoute(remoteSubnet string, operation Operation,
-			} else if currentRouteList[i].Gw.Equal(route.Gw) && currentRouteList[i].Dst.String() == route.Dst.String() {
-				klog.V(log.DEBUG).Infof("Found equivalent route, not adding")
-				found = true
-			}
-		}
-
-		if !found {
-			err = kp.netLink.RouteAdd(&route)
-			if err != nil {
-				klog.Errorf("Error adding route %s: %v", route, err)
-			}
-		}
+		desiredRoutes[key] = route
 	}
 
-	return nil
+	return desiredRoutes, nil
 }
 
-func (kp *SyncHandler) removeUnknownRoutes(vxlanGw net.IP, currentRouteList []netlink.Route) {
-	for i := range currentRouteList {
-		// Contains(endpoint destinations, route destination string, and the route gateway is our actual destination.
-		klog.V(log.DEBUG).Infof("Processing route %v", currentRouteList[i])
+// Reconcile the routes based on the GWIPs installed on this device using rtnetlink.
+func (kp *SyncHandler) reconcileIntraClusterRoutes() error {
+	currentRouteList, err := kp.listExistingRoutes()
+	if err != nil {
+		return errors.Wrapf(err, "error retrieving routes for link %s", VxLANIface)
+	}
 
-		if currentRouteList[i].Dst == nil || currentRouteList[i].Gw == nil {
-			klog.V(log.DEBUG).Infof("Found nil gw or dst")
-		} else {
-			if kp.remoteSubnets.Contains(currentRouteList[i].Dst.String()) && currentRouteList[i].Gw.Equal(vxlanGw) {
-				klog.V(log.DEBUG).Infof("Found route %s with gw %s already installed", currentRouteList[i], currentRouteList[i].Gw)
-			} else {
-				klog.V(log.DEBUG).Infof("Removing route %s", currentRouteList[i])
-				if err := kp.netLink.RouteDel(&currentRouteList[i]); err != nil {
-					klog.Errorf("Error removing route %s: %v", currentRouteList[i], err)
-				}
+	if kp.isGatewayNode {
+		klog.V(log.DEBUG).Infof("Node is a GW, delete all intra cluster Routes on %v", VxLANIface)
+		for _, route := range currentRouteList {
+			klog.V(log.DEBUG).Infof("Node is a GW Removing route %s", route.String())
+			if err = kp.netLink.RouteDel(&route); err != nil {
+				klog.Errorf("Error removing route %s: %v", route, err)
 			}
 		}
+		return nil
 	}
-}
 
-func (kp *SyncHandler) updateRoutingRulesForInterClusterSupport(remoteCIDRs []string, operation Operation) error {
-	if kp.isGatewayNode {
-		klog.V(log.DEBUG).Info("On GWNode, in updateRoutingRulesForInterClusterSupport ignoring")
-		// These rules are required only on the nonGatewayNode.
-		return nil
+	klog.V(log.DEBUG).Infof("Reconciling existing routes %v to gws %v on worker node", currentRouteList, kp.gwVTEPs.Elements())
+
+	desiredRouteList, err := kp.buildDesiredRoutes()
+	if err != nil {
+		return errors.Wrapf(err, "error retrieving desired routes for link %s", VxLANIface)
 	}
 
-	// Update internal Loadbalancing Groups
-	for _, gwIp := range kp.gwIPs.Elements() {
-		vxlanGwIP, err := getVxlanVtepIPAddress(gwIp)
-		if err != nil {
-			return errors.Wrap(err, "failed to derive the remoteVtepIP")
+	for key, route := range currentRouteList {
+		// route exists and is up to date keep it
+		if _, ok := desiredRouteList[key]; ok {
```

Comment:
The way this lookup works, if a route exists, but we're adding or removing a nexthop gateway to the Multipath field, the route won't match, and we'll end up deleting and recreating the route.  It's probably more efficient and less disruptive to do a RouteReplace instead.